File: music_.c
/**************************************************************************************************************************

    This program generates a song and writes it to a midi file.

**************************************************************************************************************************/

#include <math.h>
#include <stdio.h>
#include <stdlib.h>

#include "music_midi_write_.c"
#include "music_energy_.c"

#define	  PPQN 192

/*************************************************************************************************************************/


int main(int argc, char *argv[]){

  // files

  FILE   *pfile;
  FILE   *mfile[2];

  // general parameters

  unsigned int lcrd=96*4*4;     //in DELTAS!
  unsigned int length=96*4*64;

  // bulk parameters

  unsigned int b_channels=12;
  unsigned int b_range[16][4]={{60-24,77-24,56,127},
                               {60-12,77-12,56,127},
			       {60-12,77-12,56,127},
			       {60,77,56,127},
			       {60,77,56,127},
			       {60,77,56,127},
			       {60,77,56,127},
			       {60+12,77,56,127},
			       {60+12,77,56,127},
			       {60+24,77+24,56,127},
			       {60,77,56,127},
			       {60,77,56,127},
			       {60,77,56,127},
			       {60,77,56,127},
			       {60,77,56,127},
			       {60,77,56,127}};

  unsigned int b_gnum=2;
  double       b_gprb=0.01;

  unsigned int b_fnum=3;
  double       b_fprb=1.0/3.0;

  double       b_beta=1.25;
  double       b_wkey=1.0;
  double       b_wcrd=5.0;
  double       b_wetc=2.0;

  // sequence parameters

  unsigned int s_gnum=1;
  double       s_gprb=0.02;

  unsigned int s_channels=8;

  unsigned int s_num_min=2;
  unsigned int s_num_max=6;

  unsigned int s_nrange=12;
  unsigned int s_nroot_min=60-24;
  unsigned int s_nroot_max=60+12;

  unsigned int s_vrange=0;
  unsigned int s_vroot_min=64;
  unsigned int s_vroot_max=127;

  unsigned int s_len_min=96*4*8;
  unsigned int s_len_max=96*4*8;

  double       s_tprb=0.18;

  double       s_beta=1.25;
  double       s_wkey=1.0;
  double       s_wcrd=1.0;

  // variables

  unsigned int by[4],total[2];
  unsigned int crd,t_delta,t_now,t_prev;
  unsigned int ch,f,n,v,s,flag,tmp;
  unsigned int bulk[16],note[16];
  unsigned int seq[16],s_end[16],s_next[16],s_num[16],s_index[16],s_nroot,s_vroot,s_note[16][8],s_vel[16][8],s_time[16][8];
  double       x,y;

  // normalize parameters

  b_wkey/=(b_wkey+b_wcrd+b_wetc);
  b_wcrd/=(b_wkey+b_wcrd+b_wetc);
  b_wetc/=(b_wkey+b_wcrd+b_wetc);

  s_wkey/=(s_wkey+s_wcrd);
  s_wcrd/=(s_wkey+s_wcrd);

  // begin

  srand(time(0));

  // open file

  mfile[0]=fopen("k.b.mid","w");
  mfile[1]=fopen("k.s.mid","w");
  w_file_head(mfile[0],0,1,192);
  w_file_head(mfile[1],0,1,192);
  total[0]=w_track_head(mfile[0]);
  total[1]=w_track_head(mfile[1]);

  // set initial bulk notes

  for(ch=0;ch<b_channels;ch++){
    flag=0;
    while(flag==0){
      n=b_range[ch][0]+rand()%(b_range[ch][1]-b_range[ch][0]+1);
      y=exp(-1.0*b_beta*en_key(n));
      if(((double)rand()/(double)RAND_MAX)<y)
        flag=1;
    }
    note[ch]=n;
    bulk[ch]=rand()%2;
    if(bulk[ch]==1){
      v=b_range[ch][2]+rand()%(b_range[ch][3]-b_range[ch][2]+1);
      total[0]+=w_note_on(mfile[0],0,ch,note[ch],v);
    }
  }

  // time loop

  crd=0;
  t_prev=0;
  t_delta=0;

  for(t_now=0;t_now<length;t_now+=12){

    if(t_now-t_delta==0){

      // t_delta

      t_delta+=12*(int)pow(2,rand()%5);

      // flip chord

      if(((double)rand()/(double)RAND_MAX)<(1.0/(double)lcrd))
        crd=rand()%7;

      // flip on-off

      for(f=0;f<b_gnum;f++){
        ch=rand()%b_channels;
        x=(double)rand()/(double)RAND_MAX;
        if(x<b_gprb){
          bulk[ch]=(bulk[ch]+1)%2;
	  if(bulk[ch]==0){
	    total[0]+=w_note_off(mfile[0],t_now-t_prev,ch,note[ch],0);
	    t_prev=t_now;
 	  }
	  else{
	    v=b_range[ch][2]+rand()%(b_range[ch][3]-b_range[ch][2]+1); 
            total[0]+=w_note_on(mfile[0],t_now-t_prev,ch,note[ch],v);
	    t_prev=t_now;
	  }
        }
      }

      // flip notes

      for(f=0;f<b_fnum;f++){
        if(((double)rand()/(double)RAND_MAX)<b_fprb){
	  ch=rand()%b_channels;
	  flag=0;
	  while(flag==0){
            n=b_range[ch][0]+(rand()%(b_range[ch][1]-b_range[ch][0]+1));
            y=exp(-1.0*b_beta*(b_wkey*en_key(n)+b_wcrd*en_crd(n,crd)+b_wetc*en_etc(n,ch,b_channels,note,bulk)));
            if(((double)rand()/(double)RAND_MAX)<y)
              flag=1;
          }
	  if(bulk[ch]==1){
	    total[0]+=w_note_off(mfile[0],t_now-t_prev,ch,note[ch],0);
	    t_prev=t_now;
	    v=b_range[ch][2]+rand()%(b_range[ch][3]-b_range[ch][2]+1); 
	    total[0]+=w_note_on(mfile[0],0,ch,n,v);
  	  }
	  note[ch]=n;
        }
      }

    }

    // sequence: print

    for(ch=0;ch<s_channels;ch++){
      if(seq[ch]==1){
        if(t_now==s_end[ch]){
          seq[ch]=0;
          total[1]+=w_note_off(mfile[1],t_now-t_prev,ch,s_note[ch][s_index[ch]],0);	
	  t_prev=t_now;
        }
        else if(t_now==s_next[ch]){
          total[1]+=w_note_off(mfile[1],t_now-t_prev,ch,s_note[ch][s_index[ch]],0);
          t_prev=t_now;
	  s_index[ch]=(s_index[ch]+1)%s_num[ch];
	  total[1]+=w_note_on(mfile[1],0,ch,s_note[ch][s_index[ch]],s_vel[ch][s_index[ch]]);
	  s_next[ch]=t_now+s_time[ch][s_index[ch]];
        }
      }
    }

    // sequence: create

    for(f=0;f<s_gnum;f++){
      if(((double)rand()/(double)RAND_MAX)<s_gprb){
        ch=rand()%s_channels;
        if(seq[ch]==0){
          seq[ch]=1;
          s_end[ch]=s_len_min+rand()%(s_len_max-s_len_min+1);
          //s_end[ch]=t_now+(96*4)*(int)pow(2,rand()%4);
          s_num[ch]=s_num_min+rand()%(s_num_max-s_num_min+1);
          s_nroot=s_nroot_min+rand()%(s_nroot_max-s_nroot_min+1);
	  s_vroot=s_vroot_min+rand()%(s_vroot_max-s_vroot_min+1);
	  tmp=0;
          for(s=0;s<s_num[ch];s++){
            flag=0;
            while(flag==0){
              n=s_nroot+rand()%(s_nrange+1);
              x=(double)rand()/(double)RAND_MAX;
              y=exp(-1.0*s_beta*(s_wkey*en_key(n)+s_wcrd*en_crd(n,crd)));
              if(x<y)
                flag=1;
            }
            s_note[ch][s]=n;
	    s_vel[ch][s]=s_vroot+rand()%(s_vrange+1);
            s_time[ch][s]=24*(int)pow(2,rand()%5);
	    if(s_time[ch][s]>24 && ((double)rand()/(double)RAND_MAX)<s_tprb)
              s_time[ch][s]=(s_time[ch][s]*2)/3;
	    tmp+=s_time[ch][s];
          }
	  s_time[ch][s_num[ch]-1]=96*2*(int)ceil((double)tmp/(double)(96*2))-tmp;
          s_index[ch]=0;	    
          total[1]+=w_note_on(mfile[1],t_now-t_prev,ch,s_note[ch][s_index[ch]],s_vel[ch][s]);
          t_prev=t_now;
	  s_next[ch]=t_now+s_time[ch][s_index[ch]];
        }
      }
    }

  }

  // notes off

  for(ch=0;ch<s_channels;ch++)
    if(seq[ch]==1)
      total[1]+=w_note_off(mfile[1],0,ch,s_note[ch][s_index[ch]],0);

  total[0]+=w_note_off(mfile[0],96*4,0,note[0],0);
  for(ch=1;ch<b_channels;ch++)
    if(bulk[ch]==1)
      total[0]+=w_note_off(mfile[0],96,ch,note[ch],0);

  // fix lengths

  for(f=0;f<2;f++){
    total[f]+=w_track_end(mfile[f]);
    by[0]=total[f]/(int)pow(2,24);
    by[1]=(total[f]%(int)pow(2,24))/(int)pow(2,16);
    by[2]=(total[f]%(int)pow(2,16))/(int)pow(2,8);
    by[3]=total[f]%(int)pow(2,8);
    fseek(mfile[f],14+4,SEEK_SET);
    fputc(by[0],mfile[f]);
    fputc(by[1],mfile[f]);
    fputc(by[2],mfile[f]);
    fputc(by[3],mfile[f]);
  }

  // combine

  pfile=fopen("k.mid","w");
  w_file_head(pfile,1,2,192);
  for(f=0;f<2;f++){
    fseek(mfile[f],14,SEEK_SET);
    while(1){
      n=fgetc(mfile[f]);
      if(n!=EOF)
        fputc(n,pfile);
      else
        break;
    }
  }
  fclose(pfile);

  // close em

  fclose(mfile[0]);
  fclose(mfile[1]);

  //

  printf("\n  Writing %d bytes to k.mid\n\n",14+8+8+total[0]+total[1]);

  return;
}

File: music_.py
############################################################################################################################################

import numpy as np

import music_write_ as mid

def norm(vec):
    return [ float(v)/float(sum(vec)) for v in vec ]

############################################################################################################################################

ppqn = 192

# song parameters

song_measures = 4*4*4

beats_per_measure  = 4

chord_measures_min = 1
chord_measures_max = 6

# bulk parameters

b_num       = 10    # number of bulk notes (does not like zero, fix this later)
  
b_nroot_min = 60-24 # min pitch
b_nroot_max = 60+12 # max pitch

b_vroot_min = 32    # min velocity
b_vroot_max = 127   # max velocity

b_gnum      = 2     # number of attempted on/offs per beat
b_gprb      = 0.01  # probability of on/off

b_fnum      = 3     # number of attempted pitch changes per beat
b_fprb      = 0.333 # probability of pitch change

b_beta      = 1.25  # inverse temperature

# sequence parameters
 
s_num       = 8     # number of sequences

s_gnum      = 1     # number of attempted on/offs per step
s_gprb      = 0.02  # probability of on/off

s_imin      = 2     # min number of notes in sequence
s_imax      = 6     # max number of notes in sequence

s_nrange    = 12    # max range of pitch
s_nroot_min = 60-24 # min lowest pitch
s_nroot_max = 60+12 # max lowest pitch

s_vrange    = 0     # max range of velocity
s_vroot_min = 64    # min lowest velocity
s_vroot_max = 127   # max lowest velocity

s_beta      = 1.25   # inverse temperature

############################################################################################################################################

###
###  the 'energy' of two notes is equal to the LCD of the two frequencies
###
###    e.g., G is 3/2 the frequency of C, hence these notes have energy 2
###          E is 5/4 the frequency of C, hence these notes have energy 4
###

elist = [1., 15., 8., 5., 4., 3., 45., 2., 5., 3., 5., 8.]

def energy(m, n):
    
    return elist[abs(n - m) % 12]

###
###  the boltzmann factor of the sum of pairwise interactions is used as
###    the probability of picking a particular note
###

def factor(m, n, beta):

    e = 0.
    for i in n:
        e += energy(m, i)

    return np.exp(-beta * e)

###
###  given a key, this chooses a chord
###
###    the root is chosen uniformly from the notes in the key
###    the 3rd and 5th are then chosen by energy
###

ionian = [0, 2, 4, 5, 7, 9, 11]

def choose_chord(key, beta):
    
    a = np.random.choice(map(lambda z: z+key, ionian))
    b = np.random.choice([a+3, a+4], p = norm([factor(a+3, [key], beta), factor(a+4, [key], beta)]))
    c = np.random.choice([a+6, a+7, a+8], p = norm([factor(a+6, [key], beta), factor(a+7, [key], beta), factor(a+8, [key], beta)]))

    return map(lambda z: z%12, (a, b, c))

###
###  given a chord, this chooses a pitch to play
###

def choose_pitch(nmin, nmax, chord, beta):
    
    notes = range(nmin, nmax)    
    probs = norm(map(lambda z: factor(z, chord, beta), notes))
    
    return np.random.choice(notes, p = probs)

############################################################################################################################################

if __name__ == '__main__':
    
    # define the key

    key = 0
    
    # determine chord structure
    
    chord = []
    
    while len(chord) < song_measures:
    
        ctuple = choose_chord(key, b_beta)
        length = np.random.choice(range(chord_measures_min, chord_measures_max))
        
        for _ in range(length):  
            chord.append(ctuple)
    
    song_beats = len(chord) * beats_per_measure
    
    print 'length:', song_beats//60, 'min', song_beats%60, 'sec'

    # OPEN FILE
    
    fp  = open('k_.mid', 'wb')
    
    pos = mid.w_file_head(fp, ppqn)
    
    # TRACK ONE: bulk

    total    = mid.w_track_head(fp, beats_per_measure)
    
    t_now    = 0
    t_prev   = 0
    
    on       = [None] * b_num
    pitch    = [None] * b_num
    velocity = [None] * b_num
    
    # set initial notes

    for b in range(b_num):
        
        on[b]       = np.random.choice(range(2))
        pitch[b]    = choose_pitch(b_nroot_min, b_nroot_max, chord[0], b_beta)
        velocity[b] = np.random.choice(range(b_vroot_min, b_vroot_max))
    
        if on[b]:
            
            total += mid.w_note_on(fp, t_now-t_prev, 1, pitch[b], velocity[b])
            t_prev = t_now
            t_now += ppqn//4
    
    t_start = t_now
    
    # loop through beats
    
    for beat in range(song_beats):
    
        for b in np.random.choice(range(b_num), size=b_fnum):
            
            if np.random.random_sample() < b_fprb:
                
                if on[b]:
                    total += mid.w_note_off(fp, t_now-t_prev, 1, pitch[b], velocity[b])
                
                pitch[b]    = choose_pitch(b_nroot_min, b_nroot_max, chord[beat//beats_per_measure], b_beta)
                velocity[b] = np.random.choice(range(b_vroot_min, b_vroot_max))
                
                if on[b]:
                    total += mid.w_note_on(fp, t_now-t_prev, 1, pitch[b], velocity[b])
    
        for b in np.random.choice(range(b_num), size=b_gnum):
    
            if np.random.random_sample() < b_gprb:
            
                if not on[b]:
                    
                    on[b]  = 1
                    total += mid.w_note_on(fp, t_now-t_prev, 1, pitch[b], velocity[b])
                    t_prev = t_now
                    
                else:
                    
                    on[b]  = 0
                    total += mid.w_note_off(fp, t_now-t_prev, 1, pitch[b], velocity[b])
                    t_prev = t_now
                    
        t_now += ppqn
                    
    # turn shit off
        
    for b in range(b_num):
            
        if on[b]:
            
            on[b]  = 0
            total += mid.w_note_off(fp, t_now-t_prev, 1, pitch[b], velocity[b])
            t_prev = t_now
      
    total += mid.w_track_end(fp)
    
    mid.w_track_fix(fp, pos, total)
    
    pos += 8+total
    
    # TRACK TWO: sequences

    total    = mid.w_track_head(fp, beats_per_measure)
    
    t_now    = t_start
    t_prev   = 0
    
    on       = [None] * s_num
    ind      = [None] * s_num
    nxt      = [None] * s_num
    skip     = [None] * s_num
    pitch    = [None] * s_num
    velocity = [None] * s_num
    
    # define sequences
    
    for s in range(s_num):

        n = np.random.choice(range(s_imin, s_imax))
        r = np.random.choice(range(s_nroot_min, s_nroot_max))
        
        p = [None] * n
        v = [None] * n
    
        for i in range(n):
            
            p[i] = choose_pitch(r, r+s_nrange, [key], s_beta)
            v[i] = np.random.choice(range(s_vroot_min, s_vroot_max))
    
        on[s]       = np.random.choice(range(2))
        ind[s]      = 0
        nxt[s]      = t_now
        skip[s]     = np.random.choice([ppqn//3, ppqn//2, ppqn, ppqn*2, ppqn*3])
        pitch[s]    = p
        velocity[s] = v
        
        if on[s]:
            
            total += mid.w_note_on(fp, t_now-t_prev, s%16, pitch[s][ind[s]], velocity[s][ind[s]])
            ind[s] = (ind[s]+1)%len(pitch[s])
            nxt[s] = t_now + skip[s]
            t_prev = t_now

    # loop over divs
    
    for div in range(song_beats*ppqn):
    
        if on[s]:
            
            if t_now == nxt[s]:
                
                total += mid.w_note_off(fp, t_now-t_prev, s%16, pitch[s][ind[s]], velocity[s][ind[s]])
                t_prev = t_now
        
        if div % ppqn == 0:
        
            for s in np.random.choice(range(s_num), size=s_gnum):

                if np.random.random_sample() < s_gprb:

                    if on[s]:

                        on[s] = 0

                    else:

                        on[s]  = 1
                        ind[s] = 0
                        nxt[s] = t_now

        if on[s]:
            
            if t_now == nxt[s]:

                total += mid.w_note_on(fp, t_now-t_prev, s%16, pitch[s][ind[s]], velocity[s][ind[s]])
                ind[s] = (ind[s]+1)%len(pitch[s])
                nxt[s] = t_now + skip[s]
                t_prev = t_now
            
        t_now += 1

    # finish em
    
    for s in range(s_num):
        
        if on[s]:
            
            total += mid.w_note_off(fp, t_now-t_prev, s%12, pitch[s][ind[s]], velocity[s][ind[s]])
            t_prev = t_now

    total += mid.w_track_end(fp)

    mid.w_track_fix(fp, pos, total)
    
    pos += 8+total

    # CLOSE FILE
    
    fp.close()
            
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
            
            
            
